File: adventofcode/2023/day08/day08.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

Dest = Struct.new(:left, :right)
def parse_map(lines)
  lines.reduce({}) { |map, line|
    source, dest_tuple = line.split("=").map(&:strip)
    dests = dest_tuple.split(",").map(&:strip)
    map[source] = Dest.new dests[0][1..], dests[1][0..-2]
    map
  }
end

def part1(map, steps)
  start = 'AAA'
  steps.chars.cycle.reduce(0) {|s, c|
    break s if start == 'ZZZ'

    start = c == 'L' ? map[start].left : map[start].right
    s + 1
  }
end

def part2(map, steps)
  map
    .keys
    .select {|n| n.end_with?('A')}
    .map {|n|
      start = n
      steps.chars.cycle.reduce(0) {|s, c|
        break s if start.end_with?('Z')

        start = c == 'L' ? map[start].left : map[start].right
        s + 1
      }
    }.reduce(&:lcm)
end

lines = read_input
steps = lines.first.chomp

map = parse_map(lines[2..])
p part1(map, steps)
p part2(map, steps)

File: adventofcode/2023/day08/day08.hs
import Data.List (isSuffixOf)
import qualified Data.Map as M
import Debug.Trace (trace)

type Node = String

data Pair = Pair Node Node
    deriving (Show)

parsePair :: String -> Pair
parsePair xs = Pair left right
  where
    ps = words xs
    left = init $ tail $ head ps
    right = init $ last ps

type Map = M.Map Node Pair

parseMap :: String -> Map -> Map
parseMap [] m = m
parseMap xs m = M.insert node pair m
  where
    ps = words xs
    node = head ps
    pair = parsePair $ unwords $ tail $ tail ps

parseInput :: [String] -> (String, Map)
parseInput xs = (steps, m)
  where
    steps = head xs
    rs = tail xs
    m = foldr parseMap M.empty rs

hop :: (String, Map) -> (Node -> Bool) -> Int -> Node -> Int
hop ((x : xs), m) done count start
    | done start = count
    | x == 'L' = hop (xs ++ [x], m) done ncount left
    | otherwise = hop (xs ++ [x], m) done ncount right
  where
    ncount = count + 1
    (Just (Pair left right)) = M.lookup start m

part1 :: (String, Map) -> Int
part1 sm = hop sm done 0 "AAA"
  where
    done = (== "ZZZ")

part2 :: (String, Map) -> Int
part2 sm@(xs, m) = foldr lcm 1 counts
  where
    mkeys = M.keys m
    starts = filter (isSuffixOf "A") mkeys
    done = isSuffixOf "Z"
    counts = map (hop sm done 0) starts

main :: IO ()
main = interact $ unlines . map show . sequence [part1, part2] . parseInput . lines
